### PR TITLE
Adds text indicating the dealer should provide their business address

### PR DESCRIPTION
**Pull request by RobRuana**
Specifically the address they use for tax purposes.

This pull request also includes a bunch of other fixes regarding the dealer's business address. I'll add comments explaining each change.

### CODE REVIEW

File: uber/model_checks.py [hunk 1]
```diff
@@ -57,8 +57,25 @@ def dealer_description(group):
 
 @prereg_validation.Group
 def dealer_address(group):
-    if group.tables and not (group.address1 and group.city and group.region and group.country and group.zip_code):
```
Comment (RobRuana) [inline uber/model_checks.py]: More fine grained address validation. Derived largely from the attendee address validation found lower down in this file.

File: uber/templates/groups/form.html [hunk 2]
```diff
@@ -201,7 +201,7 @@
             <input type="text" name="website" value="{{ group.website }}" class="form-control"/>
         </div>
         <div class="col-sm-9 col-sm-offset-3">
-            <a href="{{ group.website }}">{{ group.website }}</a>
+            <a href="{% if not group.website.startswith('http') %}http://{% endif -%}{{ group.website }}">{{ group.website }}</a>
```
Comment (RobRuana) [inline uber/templates/groups/form.html]: In my testing, I was adding URLs like "www.example.com". But when I clicked the link it would take me to https://localhost:4443/uber/registration/www.example.com.

File: uber/templates/macros.html [hunk 3]
```diff
@@ -119,50 +119,51 @@
   </div>
 {%- endmacro %}
 
-{% macro address_form(group=False, attendee=False, is_prereg_dealer=False, copy_address=False, name_prefix="") -%}
+{% macro address_form(group=False, attendee=False, is_dealer=False, collect_full_address=False, copy_address=False, name_prefix="") -%}
```
Comment (RobRuana) [inline uber/templates/macros.html]: I tweaked this macro slightly. I'll try to explain each change.

I renamed `is_prereg_dealer` to `is_dealer` because we also care if it's a dealer in the admin area in addition to the preregistration pages.

File: uber/templates/macros.html [hunk 4]
```diff
@@ -119,50 +119,51 @@
...
     {% set group_or_attendee=group if group else attendee %}
     {% set id_prefix=group.id|replace("-","") if group else attendee.id|replace("-","") %}
-    {% if not is_prereg_dealer %}
+    {% if is_dealer or collect_full_address %}
```
Comment (RobRuana) [inline uber/templates/macros.html]: I think this conditional was the opposite of what we wanted. I was getting a bunch of javascript errors like `selectToAutocomplete() not a function` or something.

File: uber/templates/macros.html [hunk 5]
```diff
@@ -119,50 +119,51 @@
...
         <script src="../static/deps/selectToAutocomplete/jquery.select-to-autocomplete.js"></script>
         <script type="text/javascript">{% include "region_opts.html" %}</script>
-    {% endif %}
-    <script type="text/javascript">
```
Comment (RobRuana) [inline uber/templates/macros.html]: I just moved this inside the `{% if is_dealer or collect_full_address %}` conditional with the other full address `<script>` tags.

File: uber/templates/macros.html [hunk 6]
```diff
@@ -203,8 +204,7 @@
         </div>
     </div>
     <div class="form-group {{ id_prefix }}address_details">
-        <label for="{{ id_prefix }}address2" class="col-sm-3 control-label"></label>
```
Comment (RobRuana) [inline uber/templates/macros.html]: This empty label was throwing off the layout on small screen sizes.

File: uber/templates/groupform.html [hunk 7]
```diff
@@ -66,9 +66,6 @@
               The number of people working your table, including yourself.
             </p>
         </div>
-        <div class="form-group">
-
-        </div>
```
Comment (kitsuta) [inline uber/templates/groupform.html]: Whoops, yeah. Probably some leftover code; these files have changed a lot.

File: uber/templates/macros.html (hunk 4 above)
Comment (kitsuta) [inline uber/templates/macros.html]: That's a shame. Just to explain my thinking, the idea was "if I haven't passed the macro a variable that indicates this is the second time the macro's been called, include these scripts." Probably my issue was that I forgot to test this with collect_full_address turned off. I must have gotten all turned around.

File: uber/templates/macros.html (hunk 5 above)
Comment (kitsuta) [inline uber/templates/macros.html]: Hmm... if you were getting missing errors then clearly the old logic was no good, but won't this go the other way and result in double-printing instead? Have you tried this when `c.COLLECT_FULL_ADDRESS` is on?

File: uber/templates/preregistration/disclaimers.html [hunk 8]
```diff
@@ -9,7 +9,7 @@
     <li> MAGFest has a <a href="http://www.magfest.org/faq#-KAbb-J7_81rRy5jTDR1">one badge per person policy</a>. If you purchase multiple badges (for your family, as a gift), please ensure that each badge is purchased in the name of the individual who the badge is intended for. You may use the same email address for multiple badges so any surprises aren't ruined. </li>
     <li> Attendees (including dealers) must abide by our <a href="http://magfest.org/codeofconduct">code of conduct</a>. </li>
     {% if c.CONSENT_FORM_URL %}
-        <li> Attendees under 18 MUST bring a <a href="http://www.magfest.org/parentalconsentform">signed parental consent form</a>, which MUST be notarized if the parent is not with the minor when the badge is picked up. Also, All children 12 and under will need to be accompanied by an adult with a paid badge. </li>
+        <li> Attendees under 18 MUST bring a <a href="http://www.magfest.org/parentalconsentform">signed parental consent form</a>, which MUST be notarized if the parent is not with the minor when the badge is picked up. Also, all children 12 and under will need to be accompanied by an adult with a paid badge. </li>
```
Comment (RobRuana) [inline uber/templates/preregistration/disclaimers.html]: Haha, this capitalization change was requested by iollan on slack

File: uber/site_sections/preregistration.py [hunk 9]
```diff
@@ -113,23 +113,26 @@ def form(self, session, message='', edit_id=None, **params):
         """
         params['id'] = 'None'   # security!
 
+        # Both the Attendee class and Group class have identically named
+        # address fields. In order to distinguish the two sets of address
+        # fields in the params, the Group fields are prefixed with "group_"
+        # when the form is submitted. To prevent instantiating the Group object
+        # with the Attendee's address fields, we must clone the params and
+        # rename all the "group_" fields.
+        group_params = dict(params)
```
Comment (kitsuta) [inline uber/site_sections/preregistration.py]: Whoops, how did I miss that??

I think we do want it moved back down to after we do a group.apply -- there are some pages where there's no group_ prefix for those fields, so we want to apply the regular fields to the group first in case that's the case.

File: uber/templates/macros.html [hunk 10]
```diff
@@ -119,13 +119,13 @@
...
-{% macro address_form(group=False, attendee=False, is_prereg_dealer=False, copy_address=False, name_prefix="") -%}
+{% macro address_form(group=False, attendee=False, is_dealer=False, copy_address=False, name_prefix="") -%}
     {% set group_or_attendee=group if group else attendee %}
     {% set id_prefix=group.id|replace("-","") if group else attendee.id|replace("-","") %}
-    {% if not is_prereg_dealer %}
-        <script src="../static/deps/selectToAutocomplete/jquery.select-to-autocomplete.js"></script>
-        <script type="text/javascript">{% include "region_opts.html" %}</script>
-    {% endif %}
+
+    <script src="../static/deps/selectToAutocomplete/jquery.select-to-autocomplete.js"></script>
```
Comment (kitsuta) [inline uber/templates/macros.html]: I think this will double-print these sources if c.COLLECT_FULL_ADDRESS is turned on (that is, if the macro is called twice in a page).

You said you were getting missing script errors, so I think double-printing is the better option here and fixing that is low priority compared to making everything actually work.

File: uber/templates/macros.html (hunk 5 above)
Comment (RobRuana) [inline uber/templates/macros.html]: This PR is not finished. I removed the check for `collect_full_address` entirely.

File: uber/templates/macros.html [hunk 11]
```diff
@@ -162,7 +162,7 @@
         });
     </script>
 
-    {% if is_prereg_dealer %}
+    {% if is_dealer and not group %}
```
Comment (kitsuta) [inline uber/templates/macros.html]: Wait, this was a good change that you made... we should check c.COLLECT_FULL_ADDRESS here because the JS is useless if the full personal address isn't collected.

I don't think we need it passed to the macro cause it's a global var, but still!

Comment (RobRuana) [inline uber/templates/macros.html]: The check for collect_full_address is already done OUTSIDE the macro. If we are calling the macro, then we're gonna print the all the address fields. The check is not useful.

File: uber/templates/macros.html (hunk 10 above)
Comment (RobRuana) [inline uber/templates/macros.html]: Okay, so here's the thing, we gotta print these every time the macro is called. There's no good way to determine if they've already been included.

File: uber/templates/macros.html (hunk 11 above)
Comment (RobRuana) [inline uber/templates/macros.html]: Macros don't have access to global vars.

**Comment by RobRuana** [comment]
This whole thing is busted. Feel free to keep reviewing and leaving comments. But it's all gonna change anyway, so you may not want to spend the effort.

File: uber/site_sections/preregistration.py (hunk 9 above)
Comment (RobRuana) [inline uber/site_sections/preregistration.py]: Which pages are those? We would be better off changing those pages to use the group prefix. Having params pull double duty this way is fragile. Like... maybe these are group params, maybe not. Who knows?!

**Comment by kitsuta** [comment]
OK. :( I spent a long time trying to make the new dealer address stuff work, so it sucks that it's not there.

If you want, we can try to fix that stuff in another PR and just leave this one for the smaller changes, like the website address fix and the new explanation.

**Comment by RobRuana** [comment]
What's not there?

**Comment by kitsuta** [comment]
"There" is as in, in a good working state.

**Comment by RobRuana** [comment]
It's like 95% there!

File: uber/site_sections/preregistration.py (hunk 9 above)
Comment (kitsuta) [inline uber/site_sections/preregistration.py]: But if we change those pages, we have to also change their functions, which is a different kind of brittle... Wait, that means this block being here isn't an issue, because we WILL always use the group_ prefix in the pre-registration form. I think. I'm pretty sure.

To answer your question though, it's the group management page (group_members.html) and the admin form for groups.

File: Gruntfile.js [hunk 12]
```diff
@@ -36,8 +42,32 @@ module.exports = function (grunt) {
...
+                    'uber/static/deps/combined.min.css': ['uber/static/deps/combined.css']
+                }
+            }
         }
     });
     grunt.loadNpmTasks('grunt-bower-concat');
-    grunt.registerTask('default', ['bower_concat']);
+    grunt.loadNpmTasks('grunt-contrib-uglify');
+    grunt.loadNpmTasks('grunt-contrib-cssmin');
+    grunt.registerTask('default', ['bower_concat', 'uglify', 'cssmin']);
```
Comment (RobRuana) [inline Gruntfile.js]: This is a thing we should have been doing for a long time. I figured as long as I was adding `jquery.select-to-autocomplete.js` to `combined.js`, I might as will minify our js and css.

File: package.json [hunk 13]
```diff
@@ -2,8 +2,10 @@
...
     "version": "14.0.0",
     "devDependencies": {
-        "grunt": "0.4.5",
-        "grunt-wiredep": "2.0.0",
-        "grunt-bower-concat": "0.4.0"
+        "grunt": "1.0.1",
+        "grunt-wiredep": "3.0.1",
+        "grunt-bower-concat": "1.0.0",
+        "grunt-contrib-uglify": "3.0.1",
+        "grunt-contrib-cssmin": "2.2.0"
```
Comment (RobRuana) [inline package.json]: Updates versions, and adds minifier dependencies.

File: uber/static/theme/prereg.css [hunk 14]
```diff
@@ -1,3 +1,7 @@
+html, body {
+    min-height: 100%;
```
Comment (RobRuana) [inline uber/static/theme/prereg.css]: This is an unrelated fix for the new preregistration theme for very short forms.

File: uber/templates/macros.html [hunk 15]
```diff
@@ -119,111 +119,100 @@
...
-    {% set group_or_attendee=group if group else attendee %}
-    {% set id_prefix=group.id|replace("-","") if group else attendee.id|replace("-","") %}
-    {% if not is_prereg_dealer %}
-        <script src="../static/deps/selectToAutocomplete/jquery.select-to-autocomplete.js"></script>
-        <script type="text/javascript">{% include "region_opts.html" %}</script>
-    {% endif %}
-    <script type="text/javascript">
-        var regionChange{{ id_prefix }} = function() {
-            {{ 'setInternational();' if not group else '' }}
+{% macro address_form(model, name_prefix="", update_international=False) -%}
```
Comment (RobRuana) [inline uber/templates/macros.html]: Radically simplified address_form macro.

File: uber/templates/macros.html [hunk 16]
```diff
@@ -119,111 +119,100 @@
...
+        <label for="zip_code{{ suffix }}" class="col-sm-3 control-label">ZIP/Postal Code</label>
         <div class="col-sm-6">
-            <input type="text" name="{{ name_prefix }}zip_code" class="form-control" value="{{ group_or_attendee.zip_code }}"/>
+            <input type="text" name="{{ name_prefix }}zip_code" class="form-control" value="{{ model.zip_code }}"/>
         </div>
     </div>
 {%- endmacro %}
+
+
+{% macro toggle_checkbox(selector, label, suffix='', name='', value='1', checked=False) -%}
```
Comment (RobRuana) [inline uber/templates/macros.html]: Toggling the address fields gets it's own macro! Now we can easily add a checkbox to toggle any other elements.

File: uber/templates/preregistration/preregbase.html [hunk 17]
```diff
@@ -1,6 +1,8 @@
 {% extends "base.html" %}
 {% block head_javascript %}
     {{ super() }}
+
+    <script type="text/javascript">{% include "region_opts.html" %}</script>
```
Comment (RobRuana) [inline uber/templates/preregistration/preregbase.html]: Every preregistration page is gonna get these opts whether the page needs it or not. I don't think thats TOO big of a deal, because `region_opts.html` is actually pretty small.

File: uber/templates/macros.html (hunk 16 above)
Comment (kitsuta) [inline uber/templates/macros.html]: Nice! If we want to, in the future we could use this for the same_legal_name checkbox too. Right now that disables the field but hiding it could be just as good.

File: uber/templates/preregistration/preregbase.html (hunk 17 above)
Comment (kitsuta) [inline uber/templates/preregistration/preregbase.html]: This is fine!
